schema/query/projection_evaluator.go

```diff
@@ -0,0 +1,170 @@
+package query
+
+import (
+	"context"
+	"fmt"
+
+	"sync"
+
+	"github.com/rs/rest-layer/schema"
+)
+
+// Eval evaluate the projection on the given payload with the help of the
+// validator. The resolver is used to fetch payload of references outside of the
+// provided payload.
+func (p Projection) Eval(ctx context.Context, payload map[string]interface{}, validator schema.Validator, resolver ReferenceResolver) (map[string]interface{}, error) {
```

Comment:
> Isn't your first example's ReferenceResolver() method supposed to return a Resource interface and not a ReferenceResolver?

Not according to my original thinking on the spot, but that might be a better solution? I suppose then the interface also needs a lookup method for the Payload.